tests/racecontrol_test.py

```diff
@@ -16,79 +16,93 @@
+    del os.environ["RALLY_RUNNING_IN_DOCKER"]
 
-        self.assertEqual(expected, racecontrol.available_pipelines())
 
-    def test_prevents_running_an_unknown_pipeline(self):
-        cfg = config.Config()
-        cfg.add(config.Scope.benchmark, "race", "pipeline", "invalid")
-        cfg.add(config.Scope.benchmark, "mechanic", "distribution.version", "5.0.0")
+@pytest.fixture
+def benchmark_only_pipeline():
+    test_pipeline_name = "benchmark-only"
+    original = racecontrol.pipelines[test_pipeline_name]
+    pipeline = racecontrol.Pipeline(test_pipeline_name, "Pipeline intended for unit-testing", mock.Mock())
+    yield pipeline
+    # restore prior pipeline!
+    racecontrol.pipelines[test_pipeline_name] = original
 
-        with self.assertRaises(exceptions.SystemSetupError) as ctx:
-            racecontrol.run(cfg)
-        self.assertRegex(ctx.exception.args[0], r"Unknown pipeline \[invalid\]. List the available pipelines with [\S]+? list pipelines.")
 
-    @mock.patch.dict(os.environ, {"RALLY_RUNNING_IN_DOCKER": "true"})
-    def test_passes_benchmark_only_pipeline_in_docker(self):
-        mock_pipeline = mock.Mock()
-        test_pipeline_name = "benchmark-only"
-        racecontrol.Pipeline("benchmark-only", "Mocked benchmark-only pipeline for unittest", mock_pipeline)
-        cfg = config.Config()
-        cfg.add(config.Scope.benchmark, "race", "pipeline", "benchmark-only")
+@pytest.fixture
+def unittest_pipeline():
+    pipeline = racecontrol.Pipeline("unit-test-pipeline", "Pipeline intended for unit-testing", mock.Mock())
+    yield pipeline
+    del racecontrol.pipelines[pipeline.name]
+
+
+def test_finds_available_pipelines():
+    expected = [
+        ["from-sources", "Builds and provisions Elasticsearch, runs a benchmark and reports results."],
+        ["from-sources-complete",
+         "Builds and provisions Elasticsearch, runs a benchmark and reports results [deprecated]."],
+        ["from-sources-skip-build",
+         "Provisions Elasticsearch (skips the build), runs a benchmark and reports results [deprecated]."],
+        ["from-distribution",
+         "Downloads an Elasticsearch distribution, provisions it, runs a benchmark and reports results."],
+        ["benchmark-only", "Assumes an already running Elasticsearch instance, runs a benchmark and reports results"],
+    ]
+
+    assert expected == racecontrol.available_pipelines()
 
-        racecontrol.run(cfg)
 
-        mock_pipeline.assert_called_once_with(cfg)
+def test_prevents_running_an_unknown_pipeline():
+    cfg = config.Config()
+    cfg.add(config.Scope.benchmark, "race", "pipeline", "invalid")
+    cfg.add(config.Scope.benchmark, "mechanic", "distribution.version", "5.0.0")
 
-        del racecontrol.pipelines[test_pipeline_name]
+    with pytest.raises(exceptions.SystemSetupError) as ex:
+        racecontrol.run(cfg)
+
+    assert re.match(r"Unknown pipeline \[invalid]. List the available pipelines with [\S]+? list pipelines.",
```

Comment:
We could also use the modern feature [match](https://docs.pytest.org/en/stable/reference.html?highlight=raises#pytest.raises) feature of `pytest` and simply:

```
    with pytest.raises(
            exceptions.SystemSetupError,
            match=r"Unknown pipeline \[invalid]. List the available pipelines with [\S]+? list pipelines."):
        racecontrol.run(cfg)
```

thus removing the need to use a named resource for the exception and a separate assert (and more importantly having to know which property of the exception object holds the message :) )

Reply:
Nice, thanks for the improvement! I've pushed this in https://github.com/elastic/rally/pull/1172/commits/dff0ee0fa9b109627dfbca1beec71435337e0d19.